src/main/scala/elevate/rise/rules/algorithmic.scala

```diff
@@ -27,294 +28,204 @@ object algorithmic {
-    def apply(e: Rise): RewriteResult[Rise] = e match {
-      case App(Id() ::: FunType(in: ScalarType, out: ScalarType), arg ::: (argT: ScalarType))
-        if in == out && in == argT =>
-        Success(fun(x => x) $ arg)
-      case _ => Failure(idToCopy)
-    }
+  @rule def idToCopy: Strategy[Rise] = {
+    case App(Id() ::: FunType(in: ScalarType, out: ScalarType), arg ::: (argT: ScalarType))
+      if in == out && in == argT =>
+      Success(fun(x => x) $ arg)
   }
 
-
-  def liftId(implicit ev: Traversable[Rise]): Strategy[Rise] = `id -> *id`()
-  case class `id -> *id`()(implicit ev: Traversable[Rise]) extends Strategy[Rise] {
-    def apply(e: Rise): RewriteResult[Rise] = e match {
-      case App(Id() ::: FunType(ArrayType(_, _), _), arg) => Success(DFNF()(ev)((map(id) $ arg)).get)
-      case _ => Failure(liftId)
-    }
-    override def toString: String = "liftId"
+  @rule def liftId()(implicit ev: Traversable[Rise]): Strategy[Rise] = {
+    case App(Id() ::: FunType(ArrayType(_, _), _), arg) => Success(DFNF()(ev)((map(id) $ arg)).get)
   }
 
-  def createTransposePair: Strategy[Rise] = `id -> T >> T`
-  case object `id -> T >> T` extends Strategy[Rise] {
-    def apply(e: Rise): RewriteResult[Rise] = e match {
-      case App(Id(), arg) => Success(app(transpose >> transpose, arg) :: e.t)
-      case _              => Failure(createTransposePair)
-    }
-    override def toString: String = "createTransposePair"
+  @rule def createTransposePair: Strategy[Rise] = {
+    case e @ App(Id(), arg) => Success(app(transpose >> transpose, arg) :: e.t)
   }
 
   // _-> T >> T
   def transposePairAfter: Strategy[Rise] = idAfter `;` createTransposePair
 
-  def removeTransposePair: Strategy[Rise] = `T >> T -> `
-  case object `T >> T -> ` extends Strategy[Rise]  {
-    def apply(e: Rise): RewriteResult[Rise] = e match {
-      case App(Transpose(), App(Transpose(), x)) => Success(x :: e.t)
-      case _                                     => Failure(removeTransposePair)
-    }
-    override def toString: String = "createTransposePair"
+  @rule def removeTransposePair: Strategy[Rise] = {
+    case e @ App(Transpose(), App(Transpose(), x)) => Success(x :: e.t)
   }
 
   // overlapped tiling
 
   // constraint: n - m = u - v
   // v = u + m - n
-  def slideOverlap(u: Nat): Strategy[Rise] =
-    `slide(n, m) -> slide(u, v) >> *(slide(n, m)) >> J`(u)
-  def `slide(n, m) -> slide(u, v) >> *(slide(n, m)) >> J`(
-   u: Nat
-  ): Strategy[Rise] = {
-    case expr @ DepApp(DepApp(Slide(), n: Nat), m: Nat) =>
+  @rule def slideOverlap(u: Nat): Strategy[Rise] = {
+    case e @ DepApp(DepApp(Slide(), n: Nat), m: Nat) =>
       val v = u + m - n
-      Success((slide(u)(v) >> map(slide(n)(m)) >> join) :: expr.t)
-    case _ =>
-      Failure(slideOverlap(u))
+      Success((slide(u)(v) >> map(slide(n)(m)) >> join) :: e.t)
   }
 
   // slide widening
 
-  def dropInSlide: Strategy[Rise] =
-    `slide n 1 >> drop l -> slide (n+l) 1 >> map(drop l)`
```

Comment:
I would like to keep these kind of documentating "example/specification" as a comment.

Reply:
I added them back